omegaconf/basecontainer.py

```diff
@@ -55,25 +55,24 @@ def is_mandatory_missing(val: Any) -> bool:
 
         val = _get_value(value)
         has_default = default_value is not DEFAULT_VALUE_MARKER
-        if has_default and (val is None or is_mandatory_missing(val)):
+
+        if has_default and val is None:
             return default_value
 
+        if is_mandatory_missing(val):
+            if has_default:
+                return default_value
+            raise MissingMandatoryValue("Missing mandatory value: $FULL_KEY")
+
```

Comment:
can we simplify by combining the has_default?
```python
if has_default:
  if val is None:
    return default_value
  if is_mandatory_missing(val):
    raise ...
```

Reply:
I'm not seeing it -- the problem is to make sure that we call `is_mandatory_missing()` only once (and ideally that we don't call it at all if we don't need to). Can you provide a full snippet?

Reply:
Okay. I looked at the existing code and I am a bit confused by this change.
The current code is checking if a value is mandatory missing, if it is it returns the default if provided.
it then resolves it, and check again on the resolved value. if it's missing - it returns the default otherwise it raises.

not sure how you can consolidate the two checks into one.

Reply:
Oh, I think I understand the confusion. The reason why I could move up the check (that was after the resolution) is because it's not possible anymore for `_maybe_resolve_interpolation()` to return a node that is missing unless `val` itself is missing (since an interpolation pointing to a missing node isn't considered as missing now).

But yes, this change is a bit tricky and it's probably easier to understand by considering two scenarios to see how they play out (before / after this change):
1. `val` is missing (`"???"`)
2. `val` is an interpolation pointing to a missing node (`"${foo}"` with `foo` set to `"???"`)

In scenario 1:
* If there is a default value it is obvious that both before/after return this default value
* If there is no default value, before we would resolve `val`, which would return `"???"`, and since it is missing the `MissingMandatoryValue` exception would be raised. After this PR, this is now caught *before* resolving, and the same exception is raised.

In scenario 2:
* If there is a default value, before we would still resolve `val` (because it's an interpolation and not the missing string), which would result in a missing value, and the default value would be returned. After this PR, `resolved_node` will be `None` because the resolution will fail (due to `foo` being missing), and as a result the default value will also be returned.
* If there is no default value, before we would resolve `val`, resulting in a missing value and the `MissingMandatoryValue ` exception being thrown here. After this PR, since `throw_on_resolution_failure` is True, the resolution will fail (also with a `MissingMandatoryValue` exception)

==> in the end the behavior is the same in all situations